Add Meta support for empty_strided

ezyang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #53617 Marginally improve pytest collection for top-level test files
* #53611 Fix inaccurate dispatch table for fill_
* #53610 Convert a few more checks to raise NotImplementedError
* **#53397 Add Meta support for empty_strided**
* #53396 Add debug only layout assert for empty_cpu
* #53377 Add TORCH_CHECK_NOT_IMPLEMENTED/c10::NotImplementedError; make dispatch use it

It turns out once you remove all the indirection from the
empty_cpu_strided implementation, this implementation is pretty
simple.  We should see if we can simplify empty_cpu this way too.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D26891870](https://our.internmc.facebook.com/intern/diff/D26891870)

ezyang committed:
Add Meta support for empty_strided

It turns out once you remove all the indirection from the
empty_cpu_strided implementation, this implementation is pretty
simple.  We should see if we can simplify empty_cpu this way too.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

bdhirsh commented on aten/src/ATen/native/MetaTensor.cpp:
```diff
@@ -12,22 +12,67 @@ Tensor empty_meta(
+  // exerciseable
+  TORCH_CHECK_NOT_IMPLEMENTED(
+    layout_or_default(layout) == Layout::Strided,
+    "strided meta tensors not supported yet"
```
This is saying to error if the layout isn't strided, right? Shouldn't the error be "non-strided meta tensors are not supported yet"?

ezyang replied:
yup

ezyang committed:
Update on "Add Meta support for empty_strided"


It turns out once you remove all the indirection from the
empty_cpu_strided implementation, this implementation is pretty
simple.  We should see if we can simplify empty_cpu this way too.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D26891870](https://our.internmc.facebook.com/intern/diff/D26891870)

[ghstack-poisoned]
